diff --git a/python/ycm/vimsupport.py b/python/ycm/vimsupport.py
index 69cef18..d9933ec 100644
--- a/python/ycm/vimsupport.py
+++ b/python/ycm/vimsupport.py
@@ -267,11 +267,21 @@ class DiagnosticMatch( namedtuple( 'DiagnosticMatch',
 
 
 def GetDiagnosticMatchesInCurrentWindow():
-  vim_matches = vim.eval( 'getmatches()' )
-  return [ DiagnosticMatch( match[ 'id' ],
-                            match[ 'group' ],
-                            match[ 'pattern' ] )
-           for match in vim_matches if match[ 'group' ].startswith( 'Ycm' ) ]
+  if hasattr(vim, 'bindeval'):
+    vim_matches = vim.bindeval( 'getmatches()' )
+    vim_encoding = vim.eval('&encoding')
+    if not isinstance(vim_matches, (vim.Dictionary, vim.List)):
+      vim_matches = vim_matches.decode(vim_encoding, 'replace')
+    return [ DiagnosticMatch( match[ 'id' ].decode(vim_encoding),
+                              match[ 'group' ].decode(vim_encoding),
+                              match[ 'pattern' ].decode(vim_encoding) )
+            for match in vim_matches if match[ 'group' ].decode(vim_encoding).startswith( 'Ycm' ) ]
+  else:
+    vim_matches = vim.eval( 'getmatches()' )
+    return [ DiagnosticMatch( match[ 'id' ],
+                                match[ 'group' ],
+                                match[ 'pattern' ] )
+            for match in vim_matches if match[ 'group' ].startswith( 'Ycm' ) ]
 
 
 def AddDiagnosticMatch( match ):
